### PR TITLE
[KYUUBI #641]Support custom authentication

### PR DESCRIPTION
**Pull request by hddong**
### _Why are the changes needed?_

Users may have their own authentication system

### _How was this patch tested?_
- [X] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 1]
```diff
@@ -294,6 +294,13 @@ object KyuubiConf {
     .checkValues(AuthTypes.values.map(_.toString))
     .createWithDefault(AuthTypes.NONE.toString)
 
+  val AUTHENTICATION_CUSTOM_CLASS: OptionalConfigEntry[String] =
+    buildConf("authentication.custom.class")
+    .doc("User define authentication class.")
+    .version("1.0.0")
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:300]: 1.3.0

File: kyuubi-common/src/test/scala/org/apache/kyuubi/service/authentication/CustomAuthenticationProviderImplTest.scala [hunk 2]
```diff
@@ -0,0 +1,47 @@
+/*
```
Comment (yaooqinn) [inline kyuubi-common/src/test/scala/org/apache/kyuubi/service/authentication/CustomAuthenticationProviderImplTest.scala:1]: Can we rename to xxxSuite？

File: kyuubi-common/src/test/scala/org/apache/kyuubi/service/authentication/UserDefineAuthenticationProviderImpl.scala [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ */
+
+package org.apache.kyuubi.service.authentication
+
+import javax.security.sasl.AuthenticationException
+
+import org.apache.kyuubi.Logging
+import org.apache.kyuubi.config.KyuubiConf
+
+class UserDefineAuthenticationProviderImpl(conf: KyuubiConf)
```
Comment (yaooqinn) [inline kyuubi-common/src/test/scala/org/apache/kyuubi/service/authentication/UserDefineAuthenticationProviderImpl.scala:25]: passing `conf` here seems unnecessary?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 4]
```diff
@@ -287,13 +287,21 @@ object KyuubiConf {
...
       " <li>LDAP: Lightweight Directory Access Protocol authentication.</li></ul>")
     .version("1.0.0")
     .stringConf
     .transform(_.toUpperCase(Locale.ROOT))
     .checkValues(AuthTypes.values.map(_.toString))
     .createWithDefault(AuthTypes.NONE.toString)
 
+  val AUTHENTICATION_CUSTOM_CLASS: OptionalConfigEntry[String] =
+    buildConf("authentication.custom.class")
+    .doc("User defined authentication class.")
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:300]: User-defined authentication implementation of   org.apache.kyuubi.service.authentication.PasswdAuthenticationProvider

**Comment by yaooqinn** [comment]
Hi @hddong, thanks very much for driving this feature!

BTW, please update the `setting.md` by running tests locally on your laptop.

Also, we need to update https://kyuubi.readthedocs.io/en/latest/security/authentication.html manually according to this change.

Would you mind if we can have a section right after [using-kerberos]( https://kyuubi.readthedocs.io/en/latest/security/authentication.html#using-kerberos), e.g. `Using Custom Authentication` and give an example about how to implement and use this excellent feature?

**Comment by hddong** [comment]
@yaooqinn : thanks for your review, and of course not.

File: kyuubi-common/src/test/scala/org/apache/kyuubi/service/authentication/UserDefineAuthenticationProviderImpl.scala (hunk 3 above)
Comment (hddong) [reply kyuubi-common/src/test/scala/org/apache/kyuubi/service/authentication/UserDefineAuthenticationProviderImpl.scala:25]: > passing `conf` here seems unnecessary?

Yes, it's not used. Just want to test constructor. Will remove if needed.

Comment (yaooqinn) [reply kyuubi-common/src/test/scala/org/apache/kyuubi/service/authentication/UserDefineAuthenticationProviderImpl.scala:25]: oh, I mean can we make the `custom.class` implementation a zero-arg constructor?

Comment (hddong) [reply kyuubi-common/src/test/scala/org/apache/kyuubi/service/authentication/UserDefineAuthenticationProviderImpl.scala]: yep, had address it.

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
Thanks for your first contribution and welcome!

**Pull request closed without merging**